From 585a7fac03049a17d9412992c216ba9b44a9b003 Mon Sep 17 00:00:00 2001
From: Nicholas Bishop <nbishop@nbishop.net>
Date: Sun, 23 Aug 2026 13:56:05 -0400
Subject: [PATCH] Add wcslen builtin

In LLVM 23, calls to `wcslen` are sometimes generated by a loop optimization
pass: https://github.com/llvm/llvm-project/pull/132572. This causes some Rust
targets (including but not limited to the UEFI targets) to fail with a linker
error if a loop is transformed into a `wcslen` call. Provide a simple
implementation of `wcslen` in compiler-builtins to fix this.

Fixes https://github.com/rust-lang/rust/issues/160827
---
 .../builtins-test/tests/mem.rs                | 36 ++++++++++++++++++-
 .../compiler-builtins/src/mem/mod.rs          | 18 ++++++++++
 2 files changed, 53 insertions(+), 1 deletion(-)

diff --git a/library/compiler-builtins/builtins-test/tests/mem.rs b/library/compiler-builtins/builtins-test/tests/mem.rs
index 3c01a617950..83147840eee 100644
--- a/library/compiler-builtins/builtins-test/tests/mem.rs
+++ b/library/compiler-builtins/builtins-test/tests/mem.rs
@@ -1,5 +1,5 @@
 extern crate compiler_builtins;
-use compiler_builtins::mem::{memcmp, memcpy, memmove, memset};
+use compiler_builtins::mem::{memcmp, memcpy, memmove, memset, wchar_t};
 
 const WORD_SIZE: usize = core::mem::size_of::<usize>();
 
@@ -346,3 +346,37 @@ fn uwrite() {
         }
     }
 }
+
+/// Wrapper that prevents inlining the wcslen implementation into the test.
+///
+/// Inlining causes LLVM to pick up the wcslen pattern and generate a call to
+/// wcslen. That causes a linker error for tests on platforms without wcslen
+/// (such as wasm).
+///
+/// `no_mangle` is needed because LLVM knows not to apply the wcslen
+/// optimization to a function with this name:
+/// https://github.com/llvm/llvm-project/pull/132572
+#[inline(never)]
+#[unsafe(no_mangle)]
+unsafe fn wcslen(p: *const wchar_t) -> usize {
+    unsafe { compiler_builtins::mem::wcslen(p) }
+}
+
+#[test]
+fn test_strlen() {
+    unsafe {
+        let s = c"";
+        assert_eq!(strlen(s.as_ptr()), 0);
+        let s = c"hello, world!";
+        assert_eq!(strlen(s.as_ptr()), 13);
+    }
+}
+
+#[test]
+fn wide_string_length() {
+    let s = [0];
+    assert_eq!(unsafe { wcslen(s.as_ptr()) }, 0);
+
+    let s = [97, 98, 99, 0];
+    assert_eq!(unsafe { wcslen(s.as_ptr()) }, 3);
+}
diff --git a/library/compiler-builtins/compiler-builtins/src/mem/mod.rs b/library/compiler-builtins/compiler-builtins/src/mem/mod.rs
index ef0f4da228e..0fbb34c05cb 100644
--- a/library/compiler-builtins/compiler-builtins/src/mem/mod.rs
+++ b/library/compiler-builtins/compiler-builtins/src/mem/mod.rs
@@ -7,6 +7,13 @@
 #[cfg_attr(all(feature = "arch", target_arch = "x86_64"), path = "x86_64.rs")]
 mod impls;
 
+#[cfg(any(windows, target_os = "uefi"))]
+#[expect(non_camel_case_types)]
+pub type wchar_t = u16;
+#[cfg(not(any(windows, target_os = "uefi")))]
+#[expect(non_camel_case_types)]
+pub type wchar_t = i32;
+
 intrinsics! {
     #[mem_builtin]
     pub unsafe extern "C" fn memcpy(
@@ -67,4 +74,15 @@
     pub unsafe extern "C" fn strlen(s: *const core::ffi::c_char) -> usize {
         impls::c_string_length(s)
     }
+
+    #[mem_builtin]
+    pub unsafe extern "C" fn wcslen(s: *const crate::mem::wchar_t) -> usize {
+        let mut s = s;
+        let mut n = 0;
+        while *s != 0 {
+            n += 1;
+            s = s.add(1);
+        }
+        n
+    }
 }
-- 
2.55.0

